### ISSUE TITLE
Don't expose EC_METHOD in the public API

### ISSUE THREAD

**Issue by davidben** [openssl/openssl#6275]
`EC_METHOD` is a low-level implementation detail of the EC module, yet it is exposed in the public API. The functions in openssl/ec.h which reference it are:

* The methods themselves: `EC_GFp_simple_method`, `EC_GFp_mont_method`, etc.
* `EC_GROUP_new`
* `EC_GROUP_method_of`
* `EC_METHOD_get_field_type`
* `EC_POINT_method_of`

Note there are no APIs to create custom ones or anything (nor would it make sense to expose such a thing). Of these, `EC_GROUP_new` is just a more error-prone way of calling`EC_GROUP_new_curve_GFp`, `EC_GROUP_new_curve_GF2m`, or `EC_GROUP_new_by_curve_name`. I don't believe I've ever seen external code use it.

`EC_GROUP_method_of` and `EC_POINT_method_of` are only useful to then call `EC_METHOD_get_field_type`. Instead, we should add `EC_GROUP_get_field_type` and maybe `EC_POINT_get_field_type`. But I've never seen anyone call `EC_POINT_method_of`, so no real need for `EC_POINT_get_field_type`. (While OpenSSL does not store `EC_GROUP` references in `EC_POINT`s, one cannot write correct `EC_POINT` code without knowing the `EC_GROUP`. The security-critical point-on-curve check is only valid for the curve you checked.)

Obviously the removals can't happen until 1.2.0, but I think a clear lesson from 1.1.0 is that the new API should be added in the 1.1.x cycle.

**Comment by rwfranks** [comment]
The unwarranted presumption here is that nobody uses EC curves other than those with built-in parameters.

EC_GROUP, EC_METHOD, EC_POINT are required to set EC curve parameters and generator point from the raw numerical values.

At least one high-profile DNSSEC analysis tool would be inconvenienced if this feature disappeared anytime soon.

**Comment by davidben** [comment]
No, that is incorrect. As described above, you already can and should be using `EC_GROUP_new_curve_GFp` for that purpose. It is a much simpler and less error-prone function. (Custom curves are also a bad idea, but that's a separate issue.)

I did not suggest removing `EC_GROUP` or `EC_POINT`.

**Comment by davidben** [comment]
To that end, per #5158, 1.2.0 should also remove `EC_GROUP_set_curve_GFp` and `EC_GROUP_set_curve_GF2m` in favor of the existing `EC_GROUP_new_curve_GFp` and `EC_GROUP_new_curve_GF2m`.

**Comment by davidben** [comment]
I think you misunderstand. Again, the proposal is *not* to remove support for custom curves. It is merely to simplify the API surface for doing so, by having people use these functions, which already exist. They are vastly more convenient for the consumer and reduce complexity for OpenSSL.
https://github.com/openssl/openssl/blob/aa8f3d76fcf1502586435631be16faa1bef3cdf7/crypto/ec/ec_cvt.c

(They're not quite ideal. If we really wanted to do this right `EC_GROUP_set_generator` would be folded in as `x` and `y` parameters to an `EC_GROUP_new_curve_GFp_ex`. That an `EC_GROUP` can be in a generator-less curve-ful state in the public API causes some complexity. But at least the existing APIs already remove the curve-less state.)

**Comment by rwfranks** [comment]
On 20 May 2018 at 18:19, David Benjamin <notifications@github.com> wrote:

> I think you misunderstand. Again, the proposal is *not* to remove support
> for custom curves. It is merely to simplify the API surface for doing so,
> by having people use these functions, which already exist. They are vastly
> more convenient for the consumer and reduce complexity for OpenSSL.
> https://github.com/openssl/openssl/blob/aa8f3d76fcf1502586435631be16fa
> a1bef3cdf7/crypto/ec/ec_cvt.c
>
> Thanks, I did not realise that there already was a workable alternative in
place as far back as 1.0.1.
The offending code has been be updated and should be released in next few
weeks.
I assume the old ugliness will remain in place until the end of the 1.1.x
series to preserve your usual standard of backward compatibility.

(They're not quite ideal. If we really wanted to do this right
> EC_GROUP_set_generator would be folded in as x and y parameters to an
> EC_GROUP_new_curve_GFp_ex. That an EC_GROUP can be in a generator-less
> curve-ful state in the public API causes some complexity. But at least the
> existing APIs already remove the curve-less state.)
>
I can live with "not quite ideal" until such time as IETF gets round to
deprecating ECC-GOST (algorithm 12) when the problem disappears.

**Comment by kroeckx** [comment]
#11928 has marked all external use of EC_METHOD as deprecated.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev, barring further input

**Comment by mattcaswell** [comment]
All of these functions were deprecated in 3.0 and are eligible for removal from 4.0.

**Comment by richsalz** [comment]
The QuicTLS fix for this can be found in https://github.com/quictls/quictls/pull/155. (As of this writing, it's still glitchy, but we'll get it fixed before you get to 4.0 so it might be useful.)

**Comment by t8m** [comment]
This should definitely happen in 4.0.
